nyaa/api_torznab.py

```diff
@@ -101,71 +112,113 @@ def torznab_caps():
 
-    if season or ep or rid:
-        return render_torznab_error(201)
+    try:
+        limit = min(limit, page_size)
+        if limit == 0:
+            raise EmptySearchError('Limit of 0 guaranteed guarantees empty result')
+        if offset % limit:
+            app.logger.warn('offset {} isn\'t a multiple of the page size {}'.format(offset, limit))
+            return render_torznab_error(201)
+
+        categories = get_categories_filter(cat)
+
+        (include_tags, exclude_tags) = get_tag_filter(tag)
 
-    page = int(1 + offset / page_size)
-    # TODO: Cleanup q
-    # TODO: category can have multiple, yet search does not support that.
-    category = None
-    quality_filter = None
-    # TODO: offset = usually per page, but not guaranteed, give error if not a multiple of page size.
-    #       Or fix search module to accept offset & limit instead of page.
-    # TODO: maximum of limit should be the app.config['RESULTS_PER_PAGE']
-    query_args = {
-        'term': q or '',
-        'category': category or '0_0',
-        'quality_filter': quality_filter or '0',
-        'page': page or 1,
-        'rss': False  # We want it to page
-    }
+        # TODO: Cleanup q
 
-    # God mode
-    if flask.g.user and flask.g.user.is_admin:
-        query_args['admin'] = True
+        query_args = {
+            'term': query_term,
+            'categories': categories,
+            'include_tags': include_tags,
+            'exclude_tags': exclude_tags,
+            'page': 1 + int(offset / limit),
+            'page_size': limit
+        }
 
-    query = search(**query_args)
+        query_results = search_custom(**query_args)
+    except EmptySearchError:
+        query_results = EmptySearchResult()
 
-    return render_torznab_feed(query, offset, extended)
+    return render_torznab_feed(query_results, offset, extended)
 
 
 def handle_api_request(api_request):
-    apikey = api_request.args.get('apikey')
+    #apikey = api_request.args.get('apikey')
     mode = api_request.args.get('t')
 
     if mode == 'caps':
         return torznab_caps()
 
-    if mode == 'search':
-        return torznab_search()
+    elif mode == 'search':
+        return torznab_search(api_request)
 
-    if mode == 'tvsearch':
-        return torznab_search()
+    elif mode == 'tvsearch':
+        return torznab_search(api_request)
 
-    if mode == 'register':
+    elif mode == 'register':
         return render_torznab_error(103)
 
     # modes specified in the newznab spec that are not supported.
-    if mode in ['movie', 'music', 'book',
-                'details', 'getnfo', 'get',
-                'cartadd', 'cartdel', 'comments', 'commentadd', 'user']:
+    elif mode in ['movie', 'music', 'book',
+                  'details', 'getnfo', 'get',
+                  'cartadd', 'cartdel', 'comments', 'commentadd', 'user']:
         return render_torznab_error(203)
 
-    return render_torznab_error(202)
+    else:
+        return render_torznab_error(202)
```

Comment:
This many returns in a method is generally an antipattern.  It's better to store the result and return a single time at the end.